src/ui/content/register.tsx

```diff
@@ -1,48 +1,101 @@
-import React from 'react'
+import React, { useState } from 'react'
+import gql from 'graphql-tag'
+import { useMutation } from '@apollo/react-hooks'
 
-type MyProps = { }
-type MyState = { email: string, name: string, password: string, password2: string}
+interface UserCreateInput {
+  name: string,
+  email: string,
+  password: string,
+  age: number,
+}
 
-class RegForm extends React.Component<MyProps, MyState> {
-  constructor(props: any) {
-    super(props);
-    this.state = {email: '', name:'', password:'', password2:''};
+const REG_M = gql`
+  mutation Register($data: UserCreateInput!) {
+    signupUser(data: $data) {
+      name
+    }
+  }
+`
+function RegForm() {
+  const [email, setEmail] = useState('')
+  const [name, setName] = useState('')
+  const [password, setPassword] = useState('')
+  const [password2, setPassword2] = useState('')
 
-    this.handleChange = this.handleChange.bind(this);
-    this.handleSubmit = this.handleSubmit.bind(this);
+  //why cannot I use data:response here? is it a new "data"?
+  const registerCompleted = (data: any) => {
+    let registeredUserName:string
+    if(data &&data.signupUser) {
+      registeredUserName = data.signupUser.name
+      alert(`Successful registration for ${registeredUserName}, now you can log in.`)
+    }
   }
 
-  handleChange(event: any) {
+  const registerError = (e: any) => {
+    alert('Error in registration. Invalid data or e-mail is taken.')
+  }
+
+  const [register, { data:response, error }] = useMutation(REG_M, { errorPolicy: 'all' , onCompleted: registerCompleted, onError: registerError})
+
+  function handleSubmit(event: any) {
+    event.preventDefault()
+    if(email && email.includes('@',1) && email.includes('.',3)) {
+      if(password && password.length > 6) {
+        if(password === password2) {
+          if(name) {
+            let registerData: UserCreateInput = {name: name, email: email, password: password, age: -1}
+            register({ variables: { data: registerData } })
+              /*
```

Comment:
Here I am using the onComplete event handler, however if an error occurs, why cannot I just put register() into a try, catch block?

Reply:
see my earlier comment on imperative vs declarative. When using alerts using try catch would be easier, but for showing the error inside the form the declarative way is much more intuitive. If register would throw an error we would always have to catch which would make declarative components more complex. If you want to show an alert there is the onError callback.